lib/Doctrine/ODM/PHPCR/ReferenceManyCollection.php

```diff
@@ -100,13 +100,22 @@ public function initialize()
             $uow = $this->dm->getUnitOfWork();
             $uow->getPrefetchHelper()->prefetch($this->dm, $referencedNodes, $this->locale);
 
+            $metadata = $this->dm->getClassMetadata(get_class($this->document));
+            $metadata = $metadata->mappings[$this->property];
+            $isAssociative = $metadata['index-by-nodename'];
+
             $this->originalReferencePaths = array();
             foreach ($referencedNodes as $referencedNode) {
                 $proxy = $uow->getOrCreateProxyFromNode($referencedNode, $this->locale);
                 if (isset($targetDocument) && !$proxy instanceof $this->targetDocument) {
                     throw new PHPCRException("Unexpected class for referenced document at '{$referencedNode->getPath()}'. Expected '{$this->targetDocument}' but got '".ClassUtils::getClass($proxy)."'.");
                 }
-                $referencedDocs[] = $proxy;
+                if ($isAssociative) {
+                    $index = $referencedNode->getName();
+                    $referencedDocs[$index] = $proxy;
```

Comment:
this might overwrite other referenced documents with the same node name. we should use $referenceNode->getPath() i think.